### ISSUE TITLE
Research best way to update Nabla contracts

### ISSUE THREAD

**Issue by TorstenStueber** [pendulum-chain/wasm-deploy#36]
There are two ways to update a smart contract that has already been deployed:
* add a privileged function to the smart contract that calls the ink! function `set_code_hash()` (see [here](https://use.ink/basics/upgradeable-contracts#replacing-contract-code-with-set_code_hash)) or
* the root extrinsic `set_code` of pallet-contracts

We got the following advice:
> It does the same. The latter one is meant to be as a last resort for government to rescue a bricked contract. So the first thing is what you have to use for upgradesble contracts.

However, if we want to implement the latter, we would need to
* extend the standard Nabla contracts, e.g., by creating subcontracts that have the additional update function and deploy them instead
* find some way to make such a privileged call safe and decentralized

## TODO
Determine what is a good solution for these challenges or is it better to use the "last resort" instead?

**Comment by TorstenStueber** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Pendulum-6408a0365cda59c57d39bee5/issues/pendulum-chain/wasm-deploy/36?planning-poker) with Zenhub @b-yap @bogdanS98 @ebma @gianfra-t @TorstenStueber

**Comment by TorstenStueber** [comment]
@pendulum-chain/product this research is required before we deploy Nabla on Pendulum because we want to allow to update Nabla once it is already deployed.

**Comment by ebma** [comment]
I guess to solve this task one would need to think through both processes and compare them. But we did not define the metrics for this, how do we define the 'best' way? The approach with the least effort in terms of code changes, or in terms of general overhead (eg. in regards to setting up referenda). Anything else that matters to you here @TorstenStueber?

**Comment by TorstenStueber** [comment]
Yeah, that was quite unspecific. The statement that "The latter one is meant to be as a last resort" makes it sound like there is something bad about that approach, which I don't see – so the research would be to better understand what is bad about this approach and compare it to the downsides and effort we would need to spend to make option 1 possible.
